File: todb.rb
#encoding : utf-8

#load libs
['active_record'].each { |lib| require lib }

curr_dir = File.join(File.dirname(__FILE__), "user.db")

#Connection to sqlite3
ActiveRecord::Base.establish_connection(
	:adapter=>"sqlite3",
	:database=>"#{curr_dir}"
)

class Msg < ActiveRecord::Base
end

class AddMsg < ActiveRecord::Migration
	def self.up
		create_table :msgs do |t|
			t.string :id
			t.datetime :time
			t.string :session
			t.string :nick
			t.string :userid
			t.string :content
		end
	end

	def self.down
		drop_table :msgs
	end
end

if not Msg.table_exists?
	AddMsg.up
end

class Token < ActiveRecord::Base
end

class AddToken < ActiveRecord::Migration
	def self.up
		create_table :tokens do |t|
			t.string :id
			t.string :userid
			t.string :sessionid
			t.string :token
			t.string :flag
			t.integer :count
		end
	end

	def self.down
		drop_table :tokens
	end
end

if not Token.table_exists?
	AddToken.up
end

##### generate data list ###
class Nlist < ActiveRecord::Base
end

class AddNlist < ActiveRecord::Migration
	def self.up
		create_table :nlists do |t|
			t.string :id
		end
	end

	def self.down
		drop_table :nlists
	end
end

if not Nlist.table_exists?
	AddNlist.up
	10000.times {n = Nlist.new ; n.save}
end
############################

class MsgParse
	def initialize(file)
		@filepath = file
		@session_node_rule = /={60,}\n消息分组:(.*)\n={60,}\n消息对象:(.*)\n={60,}/
		@msg_node_rule = /(\d{4}-\d{2}-\d{2}\s*\d{1,}:\d{2}:\d{2})\s{0,}\n?(.*?)((\(|<)(.*)(\)|>)){0,1}\n(.*)/
		@count = 0
	end

	def is_session?(str)
		str =~ /={50,}/
	end

	def is_msg?(str)
		str =~ /(\d{4}-\d{2}-\d{2}\s*\d{1,}:\d{2}:\d{2})|(.*)/
	end

	def get_session(str)
		match_group = @session_node_rule.match(str)

		"#{match_group[1]} - #{match_group[2]}" unless match_group.nil?
	end

	def msg2db(str, sessionid)
	  return if sessionid.start_with? '最近联系人'
	  
		match_group = @msg_node_rule.match(str)

		unless  match_group.nil?
			#filter empty msg
			return if match_group[2] =~ /\s*\d{1,}:\d{2}:\d{2}\s*/

			msg = Msg.new
			msg.time = DateTime.strptime(match_group[1], '%Y-%m-%d %H:%M:%S')
			msg.session = sessionid
			msg.nick = match_group[2]
			msg.userid = match_group[5]
			msg.content = match_group[7]
			@count += 1
			if @count % 1000 == 0
				puts @count
			end
			return msg.save
		end
	end

	def get_n_line(file, n)
		lines = []
		n.times {lines << file.gets}
	 
		lines
	end
	
	def parse
		File.open(@filepath, :encoding=>"utf-8") do |file|
	 		curr_block = ""
			curr_node = ""

			while line = file.gets do
				next if line.index('消息记录（此消息记录为文本格式，不支持重新导入')
				if line.index("=" * 60)
					curr_node<< line << get_n_line(file, 4).join("")

					curr_block = get_session(curr_node)
					curr_node = ""
					next
				end

				if line == "\n" and curr_node.delete("\n").size != 0
					msg2db(curr_node, curr_block)
					curr_node = ""
				else
					curr_node << line
				end
			end
		end
	end
end

#file need parse...
msg = MsgParse.new("msg-test.txt")
msg.parse

File: todb.py
# encoding=utf-8

import sqlite3
import re
import sys
import datetime

# connect db
cx = sqlite3.connect('./user.db')
cu = cx.cursor()

# init database
create_msg_sql = '''
CREATE TABLE if not exists "msgs" (
    "id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, 
    "time" datetime, 
    "session" varchar, 
    "nick" varchar, 
    "userid" varchar, 
    "content" varchar
);'''

create_nlist_sql = '''
CREATE TABLE if not exists "nlists" (
    "id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL
);'''


create_token_sql = '''
CREATE TABLE if not exists "tokens" (
    "id" INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL, 
    "userid" varchar, 
    "sessionid" varchar, 
    "token" varchar, 
    "flag" varchar, 
    "count" integer
);'''

cu.execute(create_msg_sql)
cu.execute(create_nlist_sql)
cu.execute(create_token_sql)

# init nlists data
result = cu.execute("select count(*) from nlists").fetchone()
if int(result[0]) <= 0:
    cu.executemany("insert into nlists(id) values(?)", [[i] for i in range(10000)])
    cx.commit()

# parse it
msg_pattern = '(\d{4}-\d{2}-\d{2}\s*\d{1,}:\d{2}:\d{2})\s{0,}\n?(.*?)((\(|<)(.*)(\)|>)){0,1}\n(.*)'

def get_session(str):
    matches = re.match(r'={60,}\n消息分组:(.*)\n={60,}\n消息对象:(.*)\n={60,}', str)

    if matches:
        return matches.group(1) + " - " + matches.group(2)

def msg2db(str, sessionid):
    if sessionid.startswith(u'最近联系人'): return

    matches = re.match(msg_pattern, str)
    if matches:
        if re.match('.*\d{1,}:\d{2}:\d{2}\s*', matches.group(2)): return

        sql = "insert into msgs(time, session, nick, userid, content) values(?, ?, ?, ?, ?)"
        cu.execute(sql, (datetime.datetime.strptime(matches.group(1), '%Y-%m-%d %H:%M:%S'),\
                        sessionid,\
                        matches.group(2),\
                        matches.group(5),\
                        matches.group(7)))

def parse(filename):
    curr_block = ''
    curr_node  = ''
    count      = 0

    try:
        file = open(filename, 'r', encoding='utf-8')
    except IOError as e:
        print("file open error:", e)
    else:
        while True:
            line = file.readline()
            if line == '':break

            if line.find(u'消息记录（此消息记录为文本格式，不支持重新导入') >= 0:
                continue
            if line.find('=' * 60) >= 0:
                curr_node = line
                # read 4 line
                curr_node += ''.join([file.readline() for _ in range(4)])
                curr_block = get_session(curr_node)
                curr_node = ''
                continue

            if line == '\n' and len(curr_node.strip()) != 0:
                msg2db(curr_node, curr_block)
                count += 1
                if count % 1000 == 0:
                    print("parse", count, 'msgs')
                curr_node = ''
            else:
                curr_node += line

# parsing
filename = sys.argv[1]
parse(filename)
cx.commit()